src/Vault.sol

```diff
@@ -0,0 +1,489 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+import "lib/openzeppelin-contracts/contracts/token/ERC20/ERC20.sol";
+import "lib/openzeppelin-contracts/contracts/access/Ownable.sol";
+import "lib/openzeppelin-contracts/contracts/utils/math/SafeMath.sol";
+import "lib/chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
+import "./BasketHandler.sol";
+import "./PriceConverter.sol";
+import "./Coin.sol";
+import "./Portfolio.sol";
+
+/**
+ * @dev Vault contract manages a tokenized debt position.
+ *
+ * This contract provides the means for an account to manage their debt position
+ * through enforcing adequate collatoralization while withdrawing debt tokens.
+ */
+contract Vault {
+    using BasketLib for Basket;
+    using PriceConverter for Basket;
+
+    uint256 private s_debt; // In coins amount
+    bool private s_isInsolvent;
+    Basket private s_collateral;
+    address private s_user;
+    uint256 private s_lastTimeStamp;
+    address private s_notary;
+    Portfolio public s_portfolio;
+
+    Coin private immutable i_coin;
+    uint8 private immutable i_stablecoin_decimals;
+    uint256 private constant RATIO = 150;
+    uint256 private constant RATE = 5;
+    uint256 private constant PENALTY = 13;
+    uint256 private constant ERC_DECIMAL = 1e18;
+    uint256 private constant MAX_ALLOWANCE = 2 ** 256 - 1;
+
+    event CollateralAdded(address token, uint256 tokAmount);
+    event CoinMinted(address indexed debtor, uint256 debt);
+    event DebtPayed(uint256 tokAmount);
+    event VaultLiquidated();
+    event CollateralRetrieved();
+    event UserDefaulted(address indexed debtor, uint256 debt);
+    event RebalanceEvent(Portfolio.STRATEGY strategy);
+
+    modifier onlyNotary() {
+        require(
+            msg.sender == s_notary,
+            "Only the notary can call this function"
+        );
+        _;
+    }
+
+    modifier onlyUser() {
+        require(msg.sender == s_user, "Only the user can call this function");
+        _;
+    }
+
+    modifier onlyNotaryOrUser() {
+        require(msg.sender == s_user || msg.sender == s_notary);
+        _;
+    }
+
+    constructor(
+        address _coinAddress,
+        address _user,
+        address _notary,
+        address _portfolio,
+        address _EthUSD
+    ) {
```

Comment:
If we were to use the [`Ownable`](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/access/Ownable.sol) to implement the vault ownership, we would get ownership transferability out of the box.

Reply:
Agreed. I'm currently making the changes.